File: critical_functions/save_n_load/loadFPImage.m
function image = loadFPImage(file)
    %-------------------------------------------
    %   Beschreibung:
    %       loadFPImage liest FringeProcessor-Float-Dateien ein
    %       und importiert diese als MATLAB-Matrix.
    %   Parameter:
    %       file:	Name der einzulesenden Datei (string)
    %               Wird dieser leer gelassen öffnet die GUI
    %
    %   Beispiel:
    %       Image = loadFPImage('D:\Daten\Test.flt');
    %-------------------------------------------
    %
    %   edited by Martin Prinzler 2018-10-24
    %       - logical loading (fixed workaround)
    %   edited by Martin Prinzler 2016-11-08
    %       - added validation of filename
    %       - added import of filename
    %   Originial from BIAS
    
    if ~nargin
        [fileName,fileDir] = uigetfile('.fpimg','FPImage auswählen');
        file = [fileDir,fileName];
        clear fileName fileDir
    end
    
    % Datei für read-Access öffnen
    fid = fopen(file, 'r');
    
    if fid == -1
        throwAsCaller(MException('loadFPImage:fopen:file',...
            'loadFPImage couldn''t open file:\n%s',file));
    end
    
    
    % Header auswerten:
    % Aufbau des Headers:
    %   type         : IMG_TYPE                          (8 x char)
    %   headerLength : Länge des Headers                 (1 x uint16)
    %   dataOffset   : Offset der eigentlichen Daten     (1 x uint16)
    %   datatype     : Datentyp (byte, int, float, ...)  (1 x uint16)
    %   width        : Breite des Images                 (1 x uint32)
    %   height       : Höhe des Images                   (1 x uint32)
    %   depth        : Farbtiefe                         (1 x uint16)
    %   descripton   : Text                              ((dataOffset-headerLength) x char)
    magicword    = fread(fid, 8, '*char'); %#ok<NASGU>
    headerLength = fread(fid, 1, 'uint16');
    dataOffset   = fread(fid, 1, 'uint16');
    datatype     = fread(fid, 1, 'uint16');
    width        = fread(fid, 1, 'uint32');
    height       = fread(fid, 1, 'uint32');
    depth        = fread(fid, 1, 'uint16');
    descriptorLength = dataOffset - headerLength;
    descriptor   = fread(fid, descriptorLength, '*char'); %#ok<NASGU>

    %% Einlesen der Bildmatrix
    
    if datatype == 1088 && depth == 128     % complex<double>
        imageReal = fread(fid, [height, width], '*double', 8);
        fseek(fid, dataOffset + 8, 'bof');
        imageImag = fread(fid, [height, width], '*double', 8);
        image = imageReal + 1i*imageImag;
        
    elseif datatype == 64 && depth == 64    % complex<float>
        imageReal = fread(fid, [height, width], '*float', 4);
        fseek(fid, dataOffset + 4, 'bof');
        imageImag = fread(fid, [height, width], '*float', 4);
        image = imageReal + 1i*imageImag;
        
    elseif datatype == 1024 && depth == 64  % double
        image = fread(fid, [height, width], '*double');
        
    elseif datatype == 32 && depth == 32    % float
        image = fread(fid, [height, width], '*float');
        
    elseif datatype == 16 && depth == 32    % int
        image = fread(fid, [height, width], '*int32');
        
    elseif datatype == 8 && depth == 16     % short
        image = fread(fid, [height, width], '*uint16');
        
    elseif datatype == 1 && depth == 8      % byte
        image = fread(fid, [height, width], '*uint8');
        
    elseif datatype == 2 && depth == 1      % boolean (bit mask)
        %% Special bool handling
        %   make shure images are on an 8 Bit Grid
        roundedWidth = 8 * ceil(height/8);
        imageTemp = fread(fid, [roundedWidth, width], '*ubit1');
        image = imageTemp;
        %   flip order of bytes
        %   maybe the FringeProcessor is crazy?
        for k = 1:8
            image((9-k):8:roundedWidth,:) = imageTemp(k:8:roundedWidth,:);
        end
        image = logical(image(1:height,:));
        
    else
        error('Unable to read file');
    end
    
    
    image = image.';    % Transponieren
    % (weil MATLAB Zeilen und Spalten 'andersrum' indiziert als der Fringe Processor in Bildern)
    
    
    %Schließen der Datei
    status = fclose(fid); %#ok<NASGU>
    
end

File: critical_functions/save_n_load/loadFPimage.py
import numpy as np
import matplotlib.pyplot as plt
import struct
i = complex(0,1)

### decoding functions


def char(fid,num):
    result = ''
    for count in range(num):
        #x = fid.read(num)
        #for item in struct.unpack('@'+'c'*int(len(x)),x):
        x = fid.read(1)
        item = struct.unpack('@c',x)
        result += item[0].decode("utf-8") 
    return result

def bool1(fid,num):
    result = np.empty(num,dtype=bool)
    for count in range(num):
        #x = fid.read(num*1)
        #for item in struct.unpack('@?'+'c'*int(8*len(x)),x):
        x = fid.read(1)
        item = struct.unpack('@'+'?'*int(8*len(x)),x)
        for bit in range(8):
            result[count*8+bit] = item[bit]
    return result

def uint8(fid,num):
    result = np.empty(num,dtype=int)
    for count in range(num):
        #x = fid.read(num*1)
        #for item in struct.unpack('@'+'B'*int(len(x)),x):
        x = fid.read(1)
        item = struct.unpack('@B',x)
        result[count] = item[0]
    return result

def uint16(fid,num):
    result = np.empty(num,dtype=int)
    for count in range(num):
        #x = fid.read(num*2)
        #for item in struct.unpack('@'+'H'*int(len(x)/2),x):
        x = fid.read(2)
        item = struct.unpack('@H',x)
        result[count] = item[0]
    return result

def int32(fid,num):
    result = np.empty(num,dtype=int)
    for count in range(num):
        #x = fid.read(num*4)
        #for item in struct.unpack('@'+'i'*int(len(x)/4),x):
        x = fid.read(4)
        item = struct.unpack('i',x)
        result[count] = item[0]
    return result

def uint32(fid,num):
    result = np.empty(num,dtype=int)
    for count in range(num):
        #x = fid.read(num*4)
        #for item in struct.unpack('@'+'I'*int(len(x)/4),x):
        x = fid.read(4)
        item = struct.unpack('@I',x)
        result[count] = item[0]
    return result

def float(fid,num):
    result = np.empty(num,dtype='float')
    for count in range(num):
        #x = fid.read(num*4)
        #for item in struct.unpack('@'+'d'*int(len(x)/4),x):
        x = fid.read(4)
        item = struct.unpack('@d',x)
        result[count] = item[0]
    return result

def double(fid,num):
    result = np.empty(num,dtype='double')
    for count in range(num):
        #x = fid.read(num*8)
        #for item in struct.unpack('@'+'d'*int(len(x)/8),x):
        x = fid.read(8)
        item = struct.unpack('@d',x)
        result[count] = item[0]
    return result

### main function

def loadData(FileName):
    
    with open(FileName,'rb') as fid:

        magicword    = char( fid,8 )
        headerLength = uint16( fid,1 )[0]
        dataOffset   = uint16( fid,1 )[0]
        datatype     = uint16( fid,1 )[0]
        width        = uint32( fid,1 )[0]
        height       = uint32( fid,1 )[0]
        depth        = uint16( fid,1 )[0]
        descriptorLength = dataOffset - headerLength
        descriptor   = char( fid,descriptorLength )
        '''
        # print for debugging
        print( magicword )
        print( headerLength )
        print( dataOffset )
        print( datatype )
        print( width )
        print( height )
        print( depth )
        print( descriptorLength )
        print( descriptor )
        '''
        if datatype == 1088 and depth == 128:     # complex<double>
            image = double(fid,2*height*width)
            image = image[0::2] + i*image[1::2]
            image = np.reshape(image, [width,height])
        else:
            print('Unable to read file - or functionality was not implemented (edit source code to enable untested functionality.)')
        '''
        ### not tested:
        elif datatype == 64 and depth == 64:      # complex<float>
            image = float(fid,2*height*width)
            image = image[0::2] + i*image[1::2]
            image = np.reshape(image, [width,height])
        elif datatype == 1024 and depth == 64:    # double
            image = double(fid,height*width)
            image = np.reshape(image, [width,height])
        elif datatype == 32 and depth == 32:      # float
            image = float(fid,height*width)
            image = np.reshape(image, [width,height])
        elif datatype == 16 and depth == 32:      # int32
            image = int32(fid,height*width)
            image = np.reshape(image, [width,height])
        elif datatype == 8 and depth == 16:       # short
            image = uint16(fid,height*width)
            image = np.reshape(image, [width,height])
        elif datatype == 1 and depth == 8:        # byte
            image = uint8(fid,height*width)
            image = np.reshape(image, [width,height])
        elif datatype == 2 and depth == 1:        # boolean (bit mask)
            roundedWidth = 8 * np.ceil(height/8)
            image = bool1(fid,height*roundedWidth)
            for k in range(8):
                image[(8-k):roundedWidth:8,:] = image[k:roundedWidth:8,:]
            image = image[1:height,:]
        else:
            print('Unable to read file')
        '''
    return image
